bson/src/main/org/bson/internal/OverridableUuidRepresentationCodecRegistry.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.internal;
+
+import org.bson.UuidRepresentation;
+import org.bson.codecs.Codec;
+import org.bson.codecs.OverridableUuidRepresentationCodec;
+import org.bson.codecs.configuration.CodecProvider;
+
+import static org.bson.assertions.Assertions.notNull;
+
+public class OverridableUuidRepresentationCodecRegistry implements CycleDetectingCodecRegistry {
```

Comment:
This is the central abstraction that allows us to apply a `UuidRepresentation` to an existing `CodecRegistry`.

Reply:
Are there any tests proving that this is Cycle Detecting?

Reply:
There are now :)

Done.